update Helios4 BSP postinst

leggewie opened the pull request:
- drop obsolete and imperfect fix to /usr/share/initramfs-tools/hooks/mdadm
- make changes to /etc/mdadm/mdadm.conf more robust

Actually, mucking around in /etc via postinst is not allowed by policy.  We should find a better way.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Prevents unintended edits by ensuring configuration files exist before modifying RAID-related settings.

* **Refactor**
  * Streamlines how initialization steps interact with RAID setup to be more robust and conditional.

* **Chores**
  * Updates the post-install alert to reference RAID and wake-on-LAN services and notes a FIXME about installers modifying system configuration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

leggewie committed:
Helios4: drop tweak in BSP postinst that was fixed in Debian in 2020

https://bugs.debian.org/875918

truth be told, the fix in Armbian wasn't ever a proper fix as the changes
made would be overwritten by the next update of the Debian mdadm package

leggewie committed:
Helios4: make BSP postinst more robust

don't blindly assume that the mdadm package is installed

leggewie committed:
Helios4: fix incorrect build-time notification about BSP postinst

leggewie committed:
Helios4: add a FIXME warning to what we are doing via BSP postinst

iav committed:
feat(bsp/helios4): own fault-LED handling via systemd service, drop /etc/mdadm/mdadm.conf mutation

Helios4 ships a chassis red fault LED at /sys/class/leds/helios4:red:fault
that should signal mdadm array failures. Until now this was wired by having
the BSP postinst (a) append a PROGRAM /usr/sbin/mdadm-fault-led.sh line to
/etc/mdadm/mdadm.conf and (b) historically also patch
/usr/share/initramfs-tools/hooks/mdadm. The latter was dropped in the
previous commit; this one drops (a) as well and replaces it with a
self-contained systemd unit that owns the mdadm --monitor invocation.

Why this is better:

  /etc/mdadm/mdadm.conf is not a dpkg conffile but it IS generated and
  ucf-managed by the mdadm package's own postinst. Appending to it from
  another package's postinst races with ucf on every mdadm upgrade and
  is exactly the 'do not modify another package's configuration from
  maintainer scripts' anti-pattern Debian Policy Appendix E warns about.
  By driving the LED from our own service we leave mdadm.conf entirely
  alone and let mdadm own its own state.

  mdadm --monitor in daemon mode polls /proc/mdstat (default --delay=60s),
  diffs state against the previous snapshot and fork+execs the program
  passed via --program= with EVENT, MD-DEVICE and (where applicable)
  COMPONENT-DEVICE as positional args. That is exactly the contract the
  existing mdadm-fault-led.sh script was already written against, so no
  handler change is needed for this commit. ConditionPathExists= gates
  the service on the LED node actually being present, so the unit is a
  no-op on the off-chance a non-helios4 system ends up with this BSP.

Assisted-by: Claude:claude-opus-4.7

iav committed:
fix(bsp/helios4/mdadm-fault-led): light fault LED on cold-boot degraded array via NewArray probe

mdadm --monitor emits a 'NewArray' event (not 'DegradedArray') the first
time it sees each array after startup. If a disk failed while the system
was off, the array comes up already degraded, and the existing handler
silently misses it — the chassis fault LED stays dark even though the
array needs attention.

Probe state via 'mdadm --detail' on NewArray and light the fault LED if
the array reports a degraded/failed state. Ordinary clean-startup
NewArray events still leave the LED alone.

Assisted-by: Claude:claude-opus-4.7

iav committed:
refactor(bsp/helios4/mdadm-fault-led): collapse event ladder into POSIX case, fix SC2166

Replace the ladder of 'if [ $EVENT = X ]; then ... fi' with a single
'case "$EVENT" in ... esac'. The case form is the natural POSIX
construct for dispatching on a string and drops the SC2166 warning
that the previous 'if [ X = Y -o A = B ]' triggered ('[ -o ]' is
obsolescent and not well-defined under POSIX).

Drive-by:
  - quote "$TRIGGER" / "$BRIGHTNESS" expansions (best-practice, no
    behavior change — neither expands to anything with whitespace)
  - fix 'britghness' typo in the comments
  - document the positional-arg contract from mdadm --monitor

No functional change: same events produce the same LED state transitions
as before.

Assisted-by: Claude:claude-opus-4.7

iav commented:
One more nudge before we get to hardware testing — `wip-update-helios4-bsp-postinst` is currently based on a ~6-month-old `main`. Visible markers in `mvebu-helios4.inc` on this branch:

- `BOOTBRANCH='tag:v2025.10'` — `main` already moved to `v2026.04` via [`21a4d0f0b`](https://github.com/armbian/build/commit/21a4d0f0b) (helios4 u-boot bump, PR #9786).
- `UBOOT_USE_GCC='> 7.0'` — `main` dropped this variable entirely via [`c51b97321`](https://github.com/armbian/build/commit/c51b97321).

I'd like to smoke-test the `mdadm-fault-led.service` change on real helios4 hardware (have one with an actual mdraid), but doing the build against this base would validate a u-boot/toolchain combination that current Armbian users no longer get. Would you mind rebasing this branch onto current `main`? Our 3-commit follow-up stack is straightforward to re-anchor on top of your rebased branch — the three files we touch (`config/sources/families/include/mvebu-helios4.inc`, `packages/bsp/mvebu/helios4/mdadm-fault-led.sh`, the new `.service`) haven't been touched on `main` in months, so a clean apply is very likely.

No urgency on the rebase from my side; just blocking my smoke-test until the base is current.

iav committed:
fixup! feat(bsp/helios4): own fault-LED handling via systemd service, drop /etc/mdadm/mdadm.conf mutation

iav committed:
refactor(bsp/helios4/postinst): anchor PROGRAM grep, quote vars, expand ucf-aware rationale

Address CodeRabbit's nitpick on the mdadm.conf PROGRAM append:
  - Anchored grep ('^PROGRAM /usr/sbin/mdadm-fault-led.sh') matches only
    the exact directive we add, not a comment or unrelated PROGRAM line
    elsewhere in the file. The original '"PROGRAM"' substring match
    would skip the append even if the user had commented our line out.
  - Quote "$MDADM_CONF" expansions per shellcheck best-practice (no
    behavior change today — the path doesn't expand to anything with
    whitespace — but lints cleaner).
  - Single-statement 'if grep ...; then' instead of '$?'-on-next-line.
  - Drop the FIXME line: the comment block now cites the upstream
    mdmonitor.service rationale for why PROGRAM-in-mdadm.conf is the
    intended integration path (ucf preserves local edits across mdadm
    upgrades — exactly the use case ucf was designed for). Pointing to
    the upstream-documented contract is more useful than a generic
    'this is a big no-no' warning.

Smoke-tested on real helios4 mdraid hardware: ' PROGRAM /usr/sbin/mdadm-fault-led.sh' line is appended once on first postinst, idempotent on re-run; mdadm --monitor (started by upstream mdmonitor.service / mdadm.service) honors the directive and forks our handler on real Fail/DegradedArray events; chassis red fault LED lights as expected.

Reviewers — the parent commit in this branch (fixup!) is a pure revert
of the earlier '.service' approach. This commit is where the content
sits; the parent has nothing to review.

Assisted-by: Claude:claude-opus-4.7

iav committed:
feat(bsp-cli): EXTRA_BSPDEPS env-var for board-specific runtime Depends; fix helios4 to use it

armbian-bsp-cli-<board> is the natural place to express board-specific
runtime behavior — hwmon rules, services, kernel modules, and the
software it depends on. But the framework hardcodes its Depends list in
lib/functions/bsp/armbian-bsp-cli-deb.sh's reversion phase, with no
extension point for family/board configs to add board-specific deps.

PACKAGE_LIST_BOARD apt-installs packages into the rootfs at build time
but does NOT establish a dpkg-level dependency between the BSP package
and them. In a single apt transaction the order between armbian-bsp-cli-X
and an independent PACKAGE_LIST_BOARD entry like mdadm is undefined
(typically alphabetic: 'a' < 'm', so our postinst runs before mdadm's),
which races with any postinst logic that needs the dep configured first.
For deterministic ordering of postinst execution, deb-Depends is required.

Add an optional 'EXTRA_BSPDEPS' env-var read at reversion time:

  Depends: ..., device-tree-compiler${depends_base_files}${EXTRA_BSPDEPS:+, ${EXTRA_BSPDEPS}}

Default-empty, so existing boards see no behavior change. Family/board
configs that need it declare 'declare -g EXTRA_BSPDEPS="pkg1, pkg2"'
in family_tweaks_bsp() (or anywhere before reversion runs).

Use it from helios4: the previous 'local DEPENDENCIES=...; sed -i
/^Depends:/ s/...' attempt has been dead code since at least 2018 (the
sed runs in family_tweaks_bsp BEFORE the framework writes the Depends:
line, so it never matches). Replace with EXTRA_BSPDEPS="fancontrol, mdadm"
so mdadm is guaranteed installed AND configured before our BSP's postinst
appends 'PROGRAM /usr/sbin/mdadm-fault-led.sh' to /etc/mdadm/mdadm.conf.

The dead-sed hack was introduced by commit e261c6f82 (PR #1487) which
replaced a previously working 'chroot $SDCARD apt-get install fancontrol
mdadm' with the sed in DEPENDENCIES.

Assisted-by: Claude:claude-opus-4.7

iav committed:
fix(bsp/helios4/mdadm-fault-led): on RebuildFinished, keep LED on if array still degraded

The 2018 handler turned the fault LED off unconditionally on any
RebuildFinished event. But RebuildFinished is a generic 'rebuild
operation ended' transition — it covers both successful resync
(array=clean) and aborted resync (array still degraded, e.g. a freshly
added spare failed before completing). In the latter case the LED used
to go dark while the array was actually broken.

Probe array state via 'mdadm --detail' on RebuildFinished, same pattern
as the NewArray cold-boot probe added earlier in this stack. Keep the
LED solid red if state contains 'degraded' / 'FAILED'; only turn it off
on a truly clean array.

Discovered during smoke-test on real helios4 hardware: a --fail/--remove/
--add sequence on a fake md0 triggered abort (spare immediately failed
due to stale metadata), mdmonitor emitted RebuildFinished 60s later,
LED went off — but the array remained [_U] degraded.

Assisted-by: Claude:claude-opus-4.7
